### DISCUSSION TITLE
What Nuxt 3 directory should login/authentication (supabase, etc.) be handled in?

### DISCUSSION THREAD

**Discussion by silasabbott** [nuxt/nuxt#18174] [Questions]
I'm unclear of the best way to handle authentication in Nuxt 3. What is generally recognized as best practice and what folders should be used? I'm specifically using `supabase` which has both client-side and server-side.

- `/composables`
- `/plugins`
- `/server`
- `/stores` (like `pinia` or `vuex`)

**Comment by wobsoriano** [comment]
You can take inspiration from this https://github.com/wobsoriano/nuxt3-supabase but may be different from what you want to accomplish.

Basically I'd initialize the supabase client in a `plugin` and also set global user state:

```ts
// plugins/supabase.ts
export default defineNuxtPlugin((nuxtApp: NuxtApp) => {
  const supabase = createClient(supabaseUrl, supabaseKey, supabaseOptions);

  nuxtApp.provide('supabase', supabase);
  useState('user', () => supabase.auth.user())
});
```

So you can do something like this in your `layouts` or `app.vue`

```html
<script setup lang="ts">
import type { User } from '@supabase/gotrue-js';
const { $supabase } = useNuxtApp();
const user = useState<User | null>('user');

$supabase.auth.onAuthStateChange((event) => {
  if (event === 'SIGNED_IN') {
    user.value = $supabase.auth.user()
  }

  if (event === 'SIGNED_OUT') {
    user.value = null
  }
});
</script>
```

Or if you want to make your layout or `app.vue` cleaner, you can put that logic in a composable under `composables/auth.ts`

```ts
import type { User } from '@supabase/gotrue-js';

export const useAuth = () => {
  const { $supabase } = useNuxtApp();
  const user = useState<User | null>('user');

  $supabase.auth.onAuthStateChange((event) => {
    if (event === 'SIGNED_IN') {
      user.value = $supabase.auth.user()
    }

    if (event === 'SIGNED_OUT') {
      user.value = null
    }
  });
}
```

Then in your vue file

```html
<script setup lang="ts">
  useAuth()
</script>
```

**Comment by eMeRiKa13** [comment]
What about ssr? How to get the Supabase connected user from a server api file, etc?

Thank you

**Reply by eMeRiKa13** [reply]
I have my answer here https://github.com/wobsoriano/nuxt3-supabase/blob/master/packages/playground/server/api/protected.ts with your module!

**Reply by eMeRiKa13** [reply]
Thanks to your module @wobsoriano I'm trying to add my own implementation to set the Supabase cookie and get it from the server side.

Setting the Cookie seems ok. I can see in my developer console a Cookie named "sb:token", but I can't get it.

I created a server api file:
```js
import type { IncomingMessage, ServerResponse } from 'http';
import { useCookies } from 'h3';
import { createClient } from '@supabase/supabase-js/dist/main/index.js';

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore: resolved with Nuxt
import runtimeConfig from '#config';

export default async (req: IncomingMessage, res: ServerResponse) => {

  const supabase  = createClient(runtimeConfig.supabaseUrl, runtimeConfig.supabasePublicKey, null);

  // @ts-expect-error: Missing properties in h3
  req.cookies = useCookies(req);

  const user = await (await supabase.auth.api.getUserByCookie(req)).user;

  // @ts-expect-error: Missing properties in h3
  delete req.cookies;

  return user;
};
```

Called in a Pinia Store with
```js
async getServerSession() {
      try {
        await $fetch('/api/auth/get-user', {
          method: 'POST',
          credentials: 'same-origin',
        }).then(function(user) {          
          return user;
        });
      } catch (e) {
        // 404 bug
      }
}
```

but req.cookies from the server api file is empty. Do you see something I'm missing?

Thanks

**Reply by eMeRiKa13** [reply]
I made it work by sending manually the cookies as fetch option

```js
async getServerSession() {
  try {
    return await $fetch('/api/auth/get-user', {
      method: 'POST',
        headers: {
          cookie: this.$nuxt.ssrContext.req.headers['cookie']
      }
    });
  } catch (e) {
    // 404 bug
  }
},
```
